distsql: use temp storage by default

asubiotto opened the pull request:
Tested external storage on the TPC-H scalefactor 10 dataset and as discussed in last week's meeting, turning the cluster setting on by default.

Commit:
distsql: use temp storage by default

Additionally modify TestHashJoinerDrain to explicitly disable use of
temp storage because the buffer phase has to be skipped for this test to
run correctly.

rjnn reviewed (approved):
You could amend the commit message to point out that the test amendment is required because turning distsql on by default changes the test.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17665)
<!-- Reviewable:end -->

asubiotto commented:
Done. TFTR.

Commit:
distsql: fix external storage race condition

testrace failed because in external storage processors, the EvalCtx
would be copied to attach a new monitor while the possibility of it
being written to (through CollationEnvironment) still existed. Instead,
we expose a new initialization function for row containers where a
monitor can be explicitly specified, avoiding any manipulation of the
EvalCtx.

asubiotto commented:
Added a new commit that fixes the testrace failure. Please feel free to merge on my behalf if it looks good as I won't be online much over the next two weeks.

rjnn reviewed (approved):
LGTM

andreimatei commented:
Is the "distsql-disk" logictest config still useful?
https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/logic_test.go#L276

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17665#-:-Krm2yZ2wfEUhC8NK7CD:b-564682)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
In logic_test.go:682, this knob is checked, which forces DistSQL processors to spill to disk after the first byte of storage is used in their in-memory `RowContainer`s. This exercises those codepaths, which would not be exercised due to the small data sizes in our logic tests, as the default threshold is set at ~64mb (iirc).
